addons/package_hierarchy/models/package_links.py

```diff
@@ -0,0 +1,185 @@
+# -*- coding: utf-8 -*-
+
+from odoo import api, models, fields, _
+from odoo.exceptions import ValidationError
+
+
+class PackageHierarchyLink(models.Model):
+    _name = "package.hierarchy.link"
+    _description = "Links for constructing package hierarchies"
+    _order = "id"
+
+    _parent_name = "parent_id"
+
+    name = fields.Char(index=True, store=True, compute="_compute_name")
+
+    parent_id = fields.Many2one(
+        "stock.quant.package", string="Parent package", ondelete="cascade", check_company=True
+    )
+    child_id = fields.Many2one(
+        "stock.quant.package", string="Child package", ondelete="cascade", check_company=True
+    )
+    move_line_ids = fields.Many2many(
+        "stock.move.line",
+        column1="link_id",
+        column2="move_line_id",
+        string="Stock move line",
+        ondelete="cascade",
+        check_company=True,
+    )
+    company_id = fields.Many2one("res.company", default=False)
+    depth = fields.Integer(compute="_compute_would_be_depth")
+
+    def construct(self):
+        for parent_package, links in self.groupby("parent_id"):
+            links.mapped("child_id").write(
+                {"parent_id": parent_package.id if parent_package else False}
+            )
+
+    @api.model
+    def create_top_level_unlinks(self, packages, move_lines=False):
+        # Warm the cache
+        packages.get_move_lines_of_children(aux_domain=[("id", "in", move_lines.ids)])
+        link_vals = []
+        # Create unlinks from higher level packages
+        for top_level_pack in packages:
```

Comment:
Since we are not dealing with any other type of package, I think it is OK to call this variable package or pack.